dawn/prototype/mylib.hpp

```diff
@@ -274,19 +266,21 @@ class Grid {
           v.add_face(face_at(i - 1, j - 1, face_color::upward));
           v.add_face(face_at(i - 1, j - 1, face_color::downward));
         }
-        if(j > 0 || periodic) //
+        if(i < nx && j > 0 || periodic) //
           v.add_face(face_at(i, j - 1, face_color::upward));
         if(i < nx && j < ny || periodic) {
           v.add_face(face_at(i, j, face_color::downward));
           v.add_face(face_at(i, j, face_color::upward));
         }
-        v.add_face(face_at(i - 1, j, face_color::downward));
+        if(i > 0 && j < ny || periodic) {
+          v.add_face(face_at(i - 1, j, face_color::downward));
+        }
       }
   }
 
   std::vector<Face> const& faces() const { return faces_; }
   std::vector<Vertex> const& vertices() const { return vertices_; }
-  std::vector<Edge> const& edges() const { return edges_; }
+  std::vector<Edge> edges() const { return edges_; }
```

Comment:
maybe a more correct but not necessarily nicer hack would be to filter `edges_` at the end of the constructor and create a `std::vector<std::reference_wrapper<Edge>> valid_edges_`. Because now, `reduceOnEdges` might lead to wrong results if you count the number of edges. I think that should be trivial to try out, otherwise, I am ok with your solution.

Reply:
I think Lukas' solution could work. For the reason you mention, he is proposing to keep references to the original edge vector. I'll try it.

> Because now, reduceOnEdges might lead to wrong results if you count the number of edges.
Probably this is not a big issue. It would be a global reduce, which probably isn't a pattern we need to support.

Reply:
I'll try this.